File: app/assets/javascripts/cinelartv/app/models/Season.ts
import Episode, { EpisodeData } from "./Episode";

export interface SeasonData {
    id: number;
    title: string;
    description: string;
    position: number;
    episodes: EpisodeData[];
}

class Season {
    private data: SeasonData;

    constructor(data: SeasonData) {
        this.data = data;
    }

    get id(): number {
        return this.data.id;
    }

    get title(): string {
        return this.data.title;
    }

    get description(): string {
        return this.data.description;
    }

    get position(): number {
        return this.data.position;
    }

    get episodes(): Episode[] {
        return this.data.episodes.map((episodeData) => new Episode(episodeData));
    }
}

export default Season;

File: app/models/season.rb
# frozen_string_literal: true

# app/models/season.rb
class Season < ApplicationRecord
  belongs_to :content, touch: true
  has_many :episodes, dependent: :destroy

  after_commit :touch_content, on: %i[create update destroy]

  validates :title, presence: true
  validates :tmdb_id, uniqueness: true, allow_nil: true

  before_destroy :delete_episodes

  def as_json(options = {})
    super(options.merge(only: %i[id title description position tmdb_id]))
  end

  private

  def delete_episodes
    episodes.destroy_all
  end

  def touch_content
    content&.touch
  end
end
